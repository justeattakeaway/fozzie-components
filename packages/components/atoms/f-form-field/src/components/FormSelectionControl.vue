<template>
    <div
        :data-test-id="testId.container">
        <input
            :id="$attrs.id"
            :aria-labelledby="`label-${$attrs.id}`"
            v-bind="attributes"
            :type="inputType"
            :value="value"
            :data-test-id="testId.input"
            :class="[
                $style['c-formField-field'],
                $style['c-formField-field--noFocus'],
                {
                    [$style['c-formField-field--checkbox']]: isCheckbox,
                    [$style['c-formField--invalid']]: hasError
                }]"
            @change="updateSelectionControl"
        >

        <form-label
            :id="`label-${$attrs.id}`"
            :label-for="$attrs.id"
        >
            {{ labelText }}
        </form-label>
    </div>
</template>

<script>
import FormLabel from './FormLabel.vue';
import {
    VALID_SELECTION_CONTROL_TYPES,
    DEFAULT_SELECTION_CONTROL_TYPE
} from '../constants';

export default {
    components: {
        FormLabel
    },

    inheritAttrs: false,

    props: {
        labelText: {
            type: String,
            default: ''
        },
        attributes: {
            type: Object,
            default: () => ({})
        },
        inputType: {
            type: String,
            default: DEFAULT_SELECTION_CONTROL_TYPE,
            validator: value => VALID_SELECTION_CONTROL_TYPES.indexOf(value) !== -1
        },
        value: {
            type: String,
            default: ''
        },
        hasError: {
            type: Boolean,
            default: false
        }
    },

    computed: {
        testId () {
            const formFieldName = (this.attributes && this.attributes.name ? this.attributes.name : null);

            return {
                container: formFieldName ? `formfield-${formFieldName}` : 'formfield-container',
                input: formFieldName ? `formfield-${formFieldName}-${this.inputType}` : 'formfield-input'
            };
        },

        isCheckbox () {
            return this.inputType === 'checkbox';
        }
    },

    methods: {
        updateSelectionControl (event) {
            this.$emit('update', event.target.checked);
        }
    }
};
</script>

<style lang="scss" module>
    /**
        Changing the color of an inline SVG. The color needs to be encoded, so '#' turns into '%23'.
        Had a lot of problems trying to save the SVG in the `assets` folder, since it wouldn't load, and including it
        in `f-vue-icons` wouldn't help, since we need to use this in the `<style>` section, and not as a Vue component.
    */
    @mixin tick-svg($color) {
        $encodedColor: '%23' + str-slice(#{$color}, 2);
        background-image: url("data:image/svg+xml,%3C%3Fxml version='1.0' encoding='UTF-8'%3F%3E%3Csvg fill='none' viewBox='0 0 15 12' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='m4.5 9.4722-2.9861-2.9861c-0.15636-0.1569-0.36876-0.2451-0.59027-0.2451-0.22152 0-0.43392 0.0882-0.59028 0.2451-0.15691 0.15637-0.2451 0.36877-0.2451 0.59028 0 0.22152 0.088195 0.43392 0.2451 0.59028l3.5771 3.5771c0.15636 0.1569 0.36876 0.2451 0.59028 0.2451 0.22151 0 0.43391-0.0882 0.59028-0.2451l8.9229-8.9243c0.1569-0.15637 0.2451-0.36877 0.2451-0.59028 0-0.22152-0.0882-0.43392-0.2451-0.59028-0.1564-0.15691-0.3688-0.2451-0.5903-0.2451s-0.4339 0.088195-0.5903 0.2451l-8.3333 8.3333z' clip-rule='evenodd' fill='#{$encodedColor}' fill-rule='evenodd'/%3E%3C/svg%3E");
        background-repeat: no-repeat;
        background-position: center;
        background-size: 14px;
    }

    .c-formField-field--noFocus {
        &:focus,
        &:active,
        &:focus-within {
            box-shadow: none;
        }
    }

    .c-formField-field--checkbox {
        position: absolute;
        opacity: 0;
        z-index: -1;
    }

    .c-formField-field--checkbox + label {
        cursor: pointer;
        font-weight: $font-weight-regular;
        @include font-size('body-l');
    }

    .c-formField-field--checkbox:disabled + label {
        cursor: default;
    }

    .c-formField-field--checkbox + label:before {
        content: '';
        border: 1px solid $color-grey-45;
        border-radius: 2px;
        display: inline-block;
        width: 22px;
        height: 22px;
        margin-right: spacing();
        margin-top: spacing();
        vertical-align: -5px;
    }

    .c-formField-field--checkbox.c-formField--invalid + label:before {
        border-color: $form-input-borderColour--invalid;
    }

    .c-formField-field--checkbox:checked + label:before {
        @include tick-svg($color-white);

        border: 1px solid $color-interactive-brand;
        background-color: $color-interactive-brand;
    }

    .c-formField-field--checkbox:focus + label:before {
        box-shadow: 0 0 0 1px $color-white, 0 0 0 3px $color-focus;

        border: 1px solid $color-interactive-brand;

        /*
            https://stackoverflow.com/a/58570835

            Visible in Windows high-contrast themes
            box-shadow will be hidden in these modes and
            transparency will not be hidden in high-contrast
            thus box-shadow will not show but the outline will
            providing accessibility */
        outline: transparent dotted 2px;
    }

    .c-formField-field--checkbox:focus:not(:checked) + label:before {
        box-shadow: 0 0 0 1px $color-white, 0 0 0 3px $color-focus;

        border: 1px solid $color-grey-45;
    }

    .c-formField-field--checkbox:not(:disabled):not(:checked) + label:hover:before {
        background-color: darken($color-white, $color-hover-01);
    }

    .c-formField-field--checkbox:disabled + label:before {
        border: 1px solid $color-grey-30;
        background-color: $color-grey-30;
        cursor: default;
    }

    .c-formField-field--checkbox:disabled:checked + label:before {
        @include tick-svg($color-grey-45);
    }
</style>
