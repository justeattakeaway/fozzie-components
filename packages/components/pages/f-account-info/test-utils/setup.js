import Vuex from 'vuex';
import { createLocalVue } from '@vue/test-utils';
import { VueI18n } from '@justeat/f-globalisation';
import tenantConfigs from '../src/tenants';

export const localVue = createLocalVue();
localVue.use(VueI18n);
localVue.use(Vuex);

export const i18n = {
    locale: 'en-GB',
    messages: {
        'en-GB': tenantConfigs['en-GB'].messages
    }
};

export const baseUrl = 'https://smartGatewayBaseUrl.com';

export const token = 'some-auth-token';

export const conversationId = 'b7386108-95e6-4e73-9421-5b066c089153';

export const consumerDetailsGetResponse = {
    data: {
        ConsumerId: 1050450174,
        Id: '8It2lEaJjG82nJtxO00Iab0CoF0=',
        CreatedDate: '2019-04-16T17:25:00.000Z',
        EmailAddress: 'satchin.moriyoka@example.com',
        FirstName: 'Satchin',
        LastName: 'Moriyoka',
        PhoneNumber: '+441234567890',
        DateOfBirth: '1914-04-16T17:25:00.000Z',
        TermsAndConditionsId: 0,
        WantsNewsletter: true,
        AccountType: 'Default',
        ConsumerStatus: 'ReturnConsumer'
    }
};

const consumerAddressGetResponse = {
    AddressId: 1050450174,
    City: 'Bristol',
    ZipCode: 'BS1 4DJ',
    AddressName: 'Office',
    IsDefault: true, // Key field
    Line1: 'Just Eat',
    Line2: 'Broad Quay House'
};

export const consumerAddressesGetResponse = {
    data: [{
        ...consumerAddressGetResponse
    },
    {
        AddressId: 1050450175,
        City: 'Bristol',
        ZipCode: 'BS1 9DJ',
        AddressName: 'Office',
        IsDefault: false,
        Line1: 'Takeaway',
        Line2: 'Broad Quay House'
    }]
};

export const consumerDetailsMappedModel = {
    id: consumerDetailsGetResponse.data.ConsumerId,
    firstName: consumerDetailsGetResponse.data.FirstName,
    lastName: consumerDetailsGetResponse.data.LastName,
    phoneNumber: consumerDetailsGetResponse.data.PhoneNumber,
    emailAddress: consumerDetailsGetResponse.data.EmailAddress
};

export const consumerAddressMappedModel = {
    line1: consumerAddressGetResponse.Line1,
    line2: consumerAddressGetResponse.Line2,
    line3: consumerAddressGetResponse.Line3,
    locality: consumerAddressGetResponse.City,
    postcode: consumerAddressGetResponse.ZipCode
};

export const consumerViewModel = {
    ...consumerDetailsMappedModel,
    ...consumerAddressMappedModel
};

export const consumerStateModel = {
    consumer: {
        ...consumerDetailsMappedModel,
        ...consumerAddressMappedModel
    }
};
