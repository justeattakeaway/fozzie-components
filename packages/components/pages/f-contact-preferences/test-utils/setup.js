export const contactPreferencesGetResponse = {
    data: {
        Preferences: [{
            DisplayName: 'Order status',
            Sort: 1,
            Key: 'orderstatus',
            Push: false,
            Email: true,
            Sms: true
        }, {
            DisplayName: 'Review meal',
            Sort: 2,
            Key: 'reviewmeal',
            Push: false,
            Email: false,
            Sms: false
        }, {
            DisplayName: 'News & offers',
            Sort: 3,
            Key: 'news',
            Push: false,
            Email: false,
            Sms: false
        }
        ],
        PreferenceVersionViewed: 0
    }
};

export const contactPreferencesViewModel = {
    preferenceVersionViewed: 0,
    preferences: [{
        displayName: 'Order status',
        isEmailEnabled: false,
        emailValue: true,
        key: 'orderstatus',
        pushEnabled: false,
        pushValue: false,
        isSmsEnabled: false,
        smsValue: true,
        sort: 1,
        visible: true
    }, {
        displayName: 'Review meal',
        isEmailEnabled: true,
        emailValue: false,
        key: 'reviewmeal',
        pushEnabled: true,
        pushValue: false,
        isSmsEnabled: true,
        smsValue: false,
        sort: 2,
        visible: false
    }, {
        displayName: 'News & offers',
        isEmailEnabled: true,
        emailValue: false,
        key: 'news',
        pushEnabled: true,
        pushValue: false,
        isSmsEnabled: true,
        smsValue: false,
        sort: 3,
        visible: true
    }]
};

export const filteredContactPreferencesModel = [{
    displayName: 'Order status',
    isEmailEnabled: false,
    emailValue: true,
    key: 'orderstatus',
    pushEnabled: false,
    pushValue: false,
    isSmsEnabled: false,
    smsValue: true,
    sort: 1,
    visible: true
}, {
    displayName: 'News & offers',
    isEmailEnabled: true,
    emailValue: false,
    key: 'news',
    pushEnabled: true,
    pushValue: false,
    isSmsEnabled: true,
    smsValue: false,
    sort: 3,
    visible: true
}];

export const contactPreferencesUpdateModel = {
    Preferences: [{
        DisplayName: 'Order status',
        Sort: 1,
        Key: 'orderstatus',
        Push: false,
        Email: true,
        Sms: true
    }, {
        DisplayName: 'Review meal',
        Sort: 2,
        Key: 'reviewmeal',
        Push: false,
        Email: false,
        Sms: false
    }, {
        DisplayName: 'News & offers',
        Sort: 3,
        Key: 'news',
        Push: false,
        Email: false,
        Sms: false
    }],
    DeviceToken: null,
    DeviceType: null,
    PhoneNumber: null,
    PreferenceVersionViewed: 0
};

export const baseUrl = 'https://smartGatewayBaseUrl.com';

export const token = 'some-auth-token';

export const conversationId = 'b7386108-95e6-4e73-9421-5b066c089153';

export const contactPreferencesUpdateBody = {
    Preference: [{
        DisplayName: 'Order status',
        Sort: 1,
        Key: 'orderstatus',
        Push: false,
        Email: true,
        Sms: true
    }, {
        DisplayName: 'Review meal',
        Sort: 2,
        Key: 'reviewmeal',
        Push: false,
        Email: false,
        Sms: false
    }, {
        DisplayName: 'News & offers',
        Sort: 3,
        Key: 'news',
        Push: false,
        Email: true,
        Sms: false
    }
    ],
    DeviceToken: null,
    DeviceType: null,
    PhoneNumber: null,
    PreferenceVersionViewed: 0
};
