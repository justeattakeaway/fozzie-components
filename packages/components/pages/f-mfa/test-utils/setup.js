export const baseUrl = 'https://smartGatewayBaseUrl.com';

export const conversationId = 'b7386108-95e6-4e73-9421-5b066c089153';

// eslint-disable-next-line camelcase
export const postBody = { mfa_token: 'some_mfa_token', otp: 'ABC123' };
