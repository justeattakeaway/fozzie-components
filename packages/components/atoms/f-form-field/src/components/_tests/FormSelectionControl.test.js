import { shallowMount } from '@vue/test-utils';
import FormSelectionControl from '../FormSelectionControl.vue';

describe('FormSelectionControl', () => {
    let attrs;
    let propsData;

    beforeEach(() => {
        attrs = {
            disabled: null,
            id: 'uniqueId1'
        };

        propsData = {
            inputType: 'checkbox',
            value: 'testValue'
        };
    });

    it('should be defined', () => {
        const wrapper = shallowMount(FormSelectionControl, { attrs });
        expect(wrapper.exists()).toBe(true);
    });

    describe('template ::', () => {
        describe('value ::', () => {
            it('should be set using the `value` prop', () => {
                // Arrange & Act
                const wrapper = shallowMount(FormSelectionControl, { propsData, attrs });

                // Assert
                expect(wrapper.find('input').element.value).toBe(propsData.value);
            });
        });
    });

    describe('props ::', () => {
        describe('inputType ::', () => {
            it.each([
                [false, 'text'],
                [true, 'checkbox'],
                [true, 'radio']
            ])('should return %s when value is %s', (expected, inputType) => {
                // Arrange
                const { validator } = FormSelectionControl.props.inputType;

                // Act & Assert
                expect(validator(inputType)).toBe(expected);
            });
        });
    });

    describe('computed ::', () => {
        describe('testId ::', () => {
            describe('`attributes.name` is provided ::', () => {
                it('should return an object using the attribute name', async () => {
                    // Arrange
                    propsData.attributes = {
                        name: 'testCheckbox'
                    };

                    const wrapper = shallowMount(FormSelectionControl, { propsData, attrs });

                    // Act
                    const result = await wrapper.vm.testId;

                    // Assert
                    expect(result).toMatchSnapshot();
                });
            });

            describe('`attributes.name` is not provided ::', () => {
                it('should return a generic object without using the attribute name', async () => {
                    // Arrange
                    propsData.attributes = {
                        name: null
                    };

                    const wrapper = shallowMount(FormSelectionControl, { propsData, attrs });

                    // Act
                    const result = await wrapper.vm.testId;

                    // Assert
                    expect(result).toMatchSnapshot();
                });
            });
        });
    });

    describe('methods ::', () => {
        describe('updateSelectionControl ::', () => {
            it('should emit `update` when changed', async () => {
                // Arrange
                const wrapper = shallowMount(FormSelectionControl, { propsData, attrs });

                const event = {
                    target: {
                        value: 'anotherTestValue'
                    }
                };

                // Act
                await wrapper.vm.updateSelectionControl(event);

                // Assert
                expect(wrapper.emitted('update').length).toBe(1);
            });
        });
    });
});
