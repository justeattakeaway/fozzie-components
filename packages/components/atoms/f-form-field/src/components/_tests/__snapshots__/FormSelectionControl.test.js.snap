// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSelectionControl computed :: testId :: \`attributes.name\` is not provided :: should return a generic object without using the attribute name 1`] = `
Object {
  "container": "formfield-container",
  "input": "formfield-input",
}
`;

exports[`FormSelectionControl computed :: testId :: \`attributes.name\` is provided :: should return an object using the attribute name 1`] = `
Object {
  "container": "formfield-testCheckbox",
  "input": "formfield-testCheckbox-checkbox",
}
`;
